[
   {
      "author_association" : "MEMBER",
      "body" : "Thanks for your first contribution!\r\n\r\nI'm a bit divided on whether to add yet another configuration option for this. Either something like this makes sense as default behavior, or you might just as well follow the manual steps.",
      "created_at" : "2021-05-26T09:08:02Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/22072#issuecomment-848604298",
      "id" : 848604298,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/22072",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDg0ODYwNDI5OA==",
      "performed_via_github_app" : null,
      "updated_at" : "2021-05-26T09:08:02Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/848604298",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/126646?v=4",
         "events_url" : "https://api.github.com/users/laanwj/events{/privacy}",
         "followers_url" : "https://api.github.com/users/laanwj/followers",
         "following_url" : "https://api.github.com/users/laanwj/following{/other_user}",
         "gists_url" : "https://api.github.com/users/laanwj/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/laanwj",
         "id" : 126646,
         "login" : "laanwj",
         "node_id" : "MDQ6VXNlcjEyNjY0Ng==",
         "organizations_url" : "https://api.github.com/users/laanwj/orgs",
         "received_events_url" : "https://api.github.com/users/laanwj/received_events",
         "repos_url" : "https://api.github.com/users/laanwj/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/laanwj/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/laanwj/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/laanwj"
      }
   },
   {
      "author_association" : "NONE",
      "body" : "> Either something like this makes sense as default behavior, or you might just as well follow the manual steps.\r\n\r\nIn my opinion, it would make sense to have this as the default behavior, because the user can't do anything except reindexing anyway (Maybe still ask the user on the GUI though).",
      "created_at" : "2021-05-26T09:57:53Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/22072#issuecomment-848637671",
      "id" : 848637671,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/22072",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDg0ODYzNzY3MQ==",
      "performed_via_github_app" : null,
      "updated_at" : "2021-05-26T09:57:53Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/848637671",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/67546953?v=4",
         "events_url" : "https://api.github.com/users/AaronDewes/events{/privacy}",
         "followers_url" : "https://api.github.com/users/AaronDewes/followers",
         "following_url" : "https://api.github.com/users/AaronDewes/following{/other_user}",
         "gists_url" : "https://api.github.com/users/AaronDewes/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/AaronDewes",
         "id" : 67546953,
         "login" : "AaronDewes",
         "node_id" : "MDQ6VXNlcjY3NTQ2OTUz",
         "organizations_url" : "https://api.github.com/users/AaronDewes/orgs",
         "received_events_url" : "https://api.github.com/users/AaronDewes/received_events",
         "repos_url" : "https://api.github.com/users/AaronDewes/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/AaronDewes/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/AaronDewes/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/AaronDewes"
      }
   },
   {
      "author_association" : "MEMBER",
      "body" : "I'm not sure about this. If the chainstate gets corrupted the user should be made aware, so they can fix their hardware or setup.\n\nIf you reindex automatically without changing anything in the environment, there is no reason why the situation would be any different, and it'll just end up reindexing again?",
      "created_at" : "2021-05-26T16:39:24Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/22072#issuecomment-848932820",
      "id" : 848932820,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/22072",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDg0ODkzMjgyMA==",
      "performed_via_github_app" : null,
      "updated_at" : "2021-05-26T16:40:37Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/848932820",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/548488?v=4",
         "events_url" : "https://api.github.com/users/sipa/events{/privacy}",
         "followers_url" : "https://api.github.com/users/sipa/followers",
         "following_url" : "https://api.github.com/users/sipa/following{/other_user}",
         "gists_url" : "https://api.github.com/users/sipa/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/sipa",
         "id" : 548488,
         "login" : "sipa",
         "node_id" : "MDQ6VXNlcjU0ODQ4OA==",
         "organizations_url" : "https://api.github.com/users/sipa/orgs",
         "received_events_url" : "https://api.github.com/users/sipa/received_events",
         "repos_url" : "https://api.github.com/users/sipa/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/sipa/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/sipa/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/sipa"
      }
   },
   {
      "author_association" : "NONE",
      "body" : "> I'm not sure about this. If the chainstate gets corrupted the user should be made aware, so they can fix their hardware or setup.\r\n\r\nYes, so maybe leave it as an option, of course, users should fix their setup in such a case, but a sudden power outage can happen, and this would allow to try to automatically fix this.\r\n\r\n> If you reindex automatically without changing anything in the environment, there is no reason why the situation would be any different, and it'll just end up reindexing again?\r\n\r\nIf I understand the init logic correctly, the code doesn't restart from scratch if a reindex is triggered at that point, and because Bitcoin Core doesn't automatically rebuild a corrupt block database, only with this argument, I don't think such a loop is possible. Also, I'm pretty sure the part of the code suggesting a reindex is only reached when a reindex is not already running, and a reindex would most likely work.",
      "created_at" : "2021-05-26T16:56:40Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/22072#issuecomment-848945582",
      "id" : 848945582,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/22072",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDg0ODk0NTU4Mg==",
      "performed_via_github_app" : null,
      "updated_at" : "2021-05-26T16:56:40Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/848945582",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/67546953?v=4",
         "events_url" : "https://api.github.com/users/AaronDewes/events{/privacy}",
         "followers_url" : "https://api.github.com/users/AaronDewes/followers",
         "following_url" : "https://api.github.com/users/AaronDewes/following{/other_user}",
         "gists_url" : "https://api.github.com/users/AaronDewes/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/AaronDewes",
         "id" : 67546953,
         "login" : "AaronDewes",
         "node_id" : "MDQ6VXNlcjY3NTQ2OTUz",
         "organizations_url" : "https://api.github.com/users/AaronDewes/orgs",
         "received_events_url" : "https://api.github.com/users/AaronDewes/received_events",
         "repos_url" : "https://api.github.com/users/AaronDewes/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/AaronDewes/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/AaronDewes/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/AaronDewes"
      }
   },
   {
      "author_association" : "MEMBER",
      "body" : "If a power outage causes corruptions, you probably have bigger problems with your setup.\n\nMy point is: if you automatically reindex, the user isn't made aware of the corruption, can't do anything to resolve whatever is causing it, and the next time it gets corrupted you'll just start all over again.",
      "created_at" : "2021-05-26T16:59:51Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/22072#issuecomment-848947628",
      "id" : 848947628,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/22072",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDg0ODk0NzYyOA==",
      "performed_via_github_app" : null,
      "updated_at" : "2021-05-26T16:59:51Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/848947628",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/548488?v=4",
         "events_url" : "https://api.github.com/users/sipa/events{/privacy}",
         "followers_url" : "https://api.github.com/users/sipa/followers",
         "following_url" : "https://api.github.com/users/sipa/following{/other_user}",
         "gists_url" : "https://api.github.com/users/sipa/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/sipa",
         "id" : 548488,
         "login" : "sipa",
         "node_id" : "MDQ6VXNlcjU0ODQ4OA==",
         "organizations_url" : "https://api.github.com/users/sipa/orgs",
         "received_events_url" : "https://api.github.com/users/sipa/received_events",
         "repos_url" : "https://api.github.com/users/sipa/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/sipa/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/sipa/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/sipa"
      }
   },
   {
      "author_association" : "NONE",
      "body" : "> If a power outage causes corruptions, you probably have bigger problems with your setup.\r\n\r\nSolutions like RapiBlitz, Umbrel or MyNode make setting up a node really easy. These are cheap and easy to build nodes, but Raspberry Pis don't deal with power outages well, and that can often lead to data corruption. A power outage causing problems is in my opinion okay for personal nodes, many users just want to get their node set up without having to worry about (and pay for) an UPS or something similar.\r\n\r\n\r\n> My point is: if you automatically reindex, the user isn't made aware of the corruption, can't do anything to resolve whatever is causing it, and the next time it gets corrupted you'll just start all over again.\r\n\r\nShould I add a warning to the logs about it then? A sudden power outage is the most likely cause, and if something can be fixed, why not fix it automatically? A delay will be noticed by the user (and a warning), like when Bitcoin is \"Rolling forward\", which is also triggered by an unclean shutdown.",
      "created_at" : "2021-05-26T17:07:40Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/22072#issuecomment-848952572",
      "id" : 848952572,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/22072",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDg0ODk1MjU3Mg==",
      "performed_via_github_app" : null,
      "updated_at" : "2021-05-26T17:07:40Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/848952572",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/67546953?v=4",
         "events_url" : "https://api.github.com/users/AaronDewes/events{/privacy}",
         "followers_url" : "https://api.github.com/users/AaronDewes/followers",
         "following_url" : "https://api.github.com/users/AaronDewes/following{/other_user}",
         "gists_url" : "https://api.github.com/users/AaronDewes/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/AaronDewes",
         "id" : 67546953,
         "login" : "AaronDewes",
         "node_id" : "MDQ6VXNlcjY3NTQ2OTUz",
         "organizations_url" : "https://api.github.com/users/AaronDewes/orgs",
         "received_events_url" : "https://api.github.com/users/AaronDewes/received_events",
         "repos_url" : "https://api.github.com/users/AaronDewes/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/AaronDewes/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/AaronDewes/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/AaronDewes"
      }
   },
   {
      "author_association" : "MEMBER",
      "body" : "@AaronDewes Is it so easy to corrupt such hardware? That sounds terrible...\n\nI'll let others chime in about this, I'm -0 on it. I think it's always better to make sure the user is aware that corruptions happens so they can investigate - and just putting it in the log will mostly be overlooked. That said I see the convenience for setups where the user isn't directly using the node themselves.",
      "created_at" : "2021-05-26T17:15:49Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/22072#issuecomment-848960170",
      "id" : 848960170,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/22072",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDg0ODk2MDE3MA==",
      "performed_via_github_app" : null,
      "updated_at" : "2021-05-26T17:15:49Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/848960170",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/548488?v=4",
         "events_url" : "https://api.github.com/users/sipa/events{/privacy}",
         "followers_url" : "https://api.github.com/users/sipa/followers",
         "following_url" : "https://api.github.com/users/sipa/following{/other_user}",
         "gists_url" : "https://api.github.com/users/sipa/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/sipa",
         "id" : 548488,
         "login" : "sipa",
         "node_id" : "MDQ6VXNlcjU0ODQ4OA==",
         "organizations_url" : "https://api.github.com/users/sipa/orgs",
         "received_events_url" : "https://api.github.com/users/sipa/received_events",
         "repos_url" : "https://api.github.com/users/sipa/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/sipa/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/sipa/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/sipa"
      }
   },
   {
      "author_association" : "MEMBER",
      "body" : "I think provided\r\n\r\n(i) this is not the default behavior,\r\n(ii) the code change is fairly limited (as it appears to be),\r\n(iii) automatic reindexes are clearly logged, and\r\n(iv) it actually does resolve the issue of a non-UPS'd setup recovering from a power-loss based corruption,\r\n\r\nthen this is a pretty compelling feature to add.\r\n\r\nI am not sure how frequent corruption that is solely based upon power loss (vs. bad hardware) is, though.",
      "created_at" : "2021-05-26T17:21:32Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/22072#issuecomment-848963942",
      "id" : 848963942,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/22072",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDg0ODk2Mzk0Mg==",
      "performed_via_github_app" : null,
      "updated_at" : "2021-05-26T17:21:32Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/848963942",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/73197?v=4",
         "events_url" : "https://api.github.com/users/jamesob/events{/privacy}",
         "followers_url" : "https://api.github.com/users/jamesob/followers",
         "following_url" : "https://api.github.com/users/jamesob/following{/other_user}",
         "gists_url" : "https://api.github.com/users/jamesob/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/jamesob",
         "id" : 73197,
         "login" : "jamesob",
         "node_id" : "MDQ6VXNlcjczMTk3",
         "organizations_url" : "https://api.github.com/users/jamesob/orgs",
         "received_events_url" : "https://api.github.com/users/jamesob/received_events",
         "repos_url" : "https://api.github.com/users/jamesob/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/jamesob/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/jamesob/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/jamesob"
      }
   },
   {
      "author_association" : "NONE",
      "body" : "> I am not sure how frequent corruption that is solely based upon power loss (vs. bad hardware) is, though.\r\n\r\nI've been helping a lot of Umbrel users with such issues, and if it was a bad drive, either\r\n\r\n(a) The drive wouldn't mount\r\n(b) There would be different errors (that I don't remember right now)\r\n\r\nEvery time this message got printed, a reindex actually helped.\r\n\r\n> and just putting it in the log will mostly be overlooked.\r\n\r\nFor this, it maybe could help to provide information that the node is reindexing over the RPC api, becauce I can't find such an endpoint. This way, custom UIs build on top of that API could also prompt the user with a warning, so it is more likely to be seen. I'm not sure if that should be part of this or another PR though.\r\n\r\n> (iii) automatic reindexes are clearly logged\r\n\r\nI'll do that.\r\n\r\n> (iv) it actually does resolve the issue of a non-UPS'd setup recovering from a power-loss based corruption,\r\n\r\nA manual reindex did it, so it should be the same for this.",
      "created_at" : "2021-05-26T17:27:32Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/22072#issuecomment-848970421",
      "id" : 848970421,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/22072",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDg0ODk3MDQyMQ==",
      "performed_via_github_app" : null,
      "updated_at" : "2021-05-26T17:27:32Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/848970421",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/67546953?v=4",
         "events_url" : "https://api.github.com/users/AaronDewes/events{/privacy}",
         "followers_url" : "https://api.github.com/users/AaronDewes/followers",
         "following_url" : "https://api.github.com/users/AaronDewes/following{/other_user}",
         "gists_url" : "https://api.github.com/users/AaronDewes/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/AaronDewes",
         "id" : 67546953,
         "login" : "AaronDewes",
         "node_id" : "MDQ6VXNlcjY3NTQ2OTUz",
         "organizations_url" : "https://api.github.com/users/AaronDewes/orgs",
         "received_events_url" : "https://api.github.com/users/AaronDewes/received_events",
         "repos_url" : "https://api.github.com/users/AaronDewes/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/AaronDewes/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/AaronDewes/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/AaronDewes"
      }
   },
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "This seems like a good feature. Just to bikeshed, maybe this could be a new reindex value, `-reindex=auto` complementing current `-reindex=0` and `-reindex=1` so it is less vague than `-attemptrecovery` and maybe more discoverable. `attemptrecovery` sounds more like a one time emergency flag to me than something I would configure permanently on my raspberry pi.",
      "created_at" : "2021-05-26T18:25:37Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/22072#issuecomment-849017244",
      "id" : 849017244,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/22072",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDg0OTAxNzI0NA==",
      "performed_via_github_app" : null,
      "updated_at" : "2021-05-26T18:25:53Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/849017244",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/7133040?v=4",
         "events_url" : "https://api.github.com/users/ryanofsky/events{/privacy}",
         "followers_url" : "https://api.github.com/users/ryanofsky/followers",
         "following_url" : "https://api.github.com/users/ryanofsky/following{/other_user}",
         "gists_url" : "https://api.github.com/users/ryanofsky/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/ryanofsky",
         "id" : 7133040,
         "login" : "ryanofsky",
         "node_id" : "MDQ6VXNlcjcxMzMwNDA=",
         "organizations_url" : "https://api.github.com/users/ryanofsky/orgs",
         "received_events_url" : "https://api.github.com/users/ryanofsky/received_events",
         "repos_url" : "https://api.github.com/users/ryanofsky/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/ryanofsky/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/ryanofsky/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/ryanofsky"
      }
   }
]
